indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexPhaseRunner.java

```diff
@@ -0,0 +1,467 @@
+                        monitorStatus.getExpectedSucceeded(),
+                        monitorStatus.getSucceeded(),
+                        monitorStatus.getFailed()
+                    );
+                  }
+                }
+              } else if (taskMonitor.getNumRunningTasks() < maxNumTasks) {
+                // We have more subTasks to run
+                submitNewTask(taskMonitor, subTaskSpecIterator.next());
+              } else {
+                // We have more subTasks to run, but don't have enough available task slots
+                // do nothing
+              }
+              break;
+            case FAILED:
+              // TaskMonitor already tried everything it can do for failed tasks. We failed.
+              state = TaskState.FAILED;
+              subTaskScheduleAndMonitorStopped = true;
+              final TaskStatusPlus lastStatus = taskCompleteEvent.getLastStatus();
+              if (lastStatus != null) {
+                LOG.error("Failed because of the failed sub task[%s]", lastStatus.getId());
+              } else {
+                final SinglePhaseSubTaskSpec spec =
+                    (SinglePhaseSubTaskSpec) taskCompleteEvent.getSpec();
+                LOG.error(
+                    "Failed to run sub tasks for inputSplits[%s]",
+                    getSplitsIfSplittable(spec.getIngestionSpec().getIOConfig().getFirehoseFactory())
+                );
+              }
+              break;
+            default:
+              throw new ISE("spec[%s] is in an invalid state[%s]", taskCompleteEvent.getSpec().getId(), completeState);
+          }
+        }
+      }
+    }
+    finally {
+      stopInternal();
+      if (!state.isComplete()) {
+        state = TaskState.FAILED;
+      }
+    }
+
+    return state;
+  }
+
+  private boolean isRunning()
+  {
+    return !subTaskScheduleAndMonitorStopped && !Thread.currentThread().isInterrupted();
+  }
+
+  private void submitNewTask(
+      TaskMonitor<SubTaskType> taskMonitor,
+      SubTaskSpec<SubTaskType> spec
+  )
+  {
+    LOG.info("Submit a new task for spec[%s] and inputSplit[%s]", spec.getId(), spec.getInputSplit());
+    final ListenableFuture<SubTaskCompleteEvent<SubTaskType>> future = taskMonitor.submit(spec);
+    Futures.addCallback(
+        future,
+        new FutureCallback<SubTaskCompleteEvent<SubTaskType>>()
+        {
+          @Override
+          public void onSuccess(SubTaskCompleteEvent<SubTaskType> completeEvent)
+          {
+            // this callback is called if a task completed wheter it succeeded or not.
+            taskCompleteEvents.offer(completeEvent);
+          }
+
+          @Override
+          public void onFailure(Throwable t)
+          {
+            // this callback is called only when there were some problems in TaskMonitor.
+            LOG.error(t, "Error while running a task for subTaskSpec[%s]", spec);
```

Comment:
Suggest revising the error log message to mention that it indicates issues with TaskMonitor specifically

Reply:
Would you elaborate more on what you think should be mentioned? The `Future` returned from `taskMonitor.submit()` indicates the result of processing a sub task spec after a certain amount of task retries on failures.

Reply:
I think the comments here could discuss how `onFailure` is triggered after task retries are exhausted, and `onSuccess` is called for individual task success

Reply:
Added a javadoc to TaskMonitor.submit().